### ISSUE TITLE
background img_tiles stretched along x axis with custom Mercator projection

### ISSUE THREAD

**Issue by fabio1299** [SciTools/cartopy#1055]
### Description

I'm defining custom Mercator projections to map countries that span across the 180/-180 meridian.
I use img_tiles to add a background image to my map. The country boundaries are loaded from a shapefile using geopandas.

When I use zoom levels of either 0 or 1 everything works fine, but when I use higher levels of zoom the background image starts getting distorted (stretched along the x axis away from the 180/-180 meridian). As the zoom level increases the stretch of the background image also increases.
However this only happens when an extent is set for the axes by either using the .set_extent method or by plotting selected country with the geodataframe method .plot

The following are examples of the results with zoom levels of 1,2 and 3 respectively:
<img width="500" alt="screen shot 2018-03-17 at 3 24 21 pm" src="https://user-images.githubusercontent.com/11065764/37559370-17f03fa4-29fb-11e8-8c43-b6792b127d40.png">
<img width="499" alt="screen shot 2018-03-17 at 3 25 01 pm" src="https://user-images.githubusercontent.com/11065764/37559371-1afa90a0-29fb-11e8-913f-e25e3b34c085.png">
<img width="500" alt="screen shot 2018-03-17 at 3 25 40 pm" src="https://user-images.githubusercontent.com/11065764/37559372-1e072ae2-29fb-11e8-8ce1-9ecb7161b3f2.png">

#### Code to reproduce
```
 from matplotlib import pyplot as plt
 import geopandas as gp
 import cartopy.crs as ccrs
 import cartopy.io.img_tiles as cimgt
 import cartopy.feature as cfeature
 import os

 def doPlot():

    # We get the same result with two different types of background
    # image tiles (e.g. Stamen Terrain and OpenStreetMap) 
    # We haven't tested any other backgrounds
    background_tiles = cimgt.StamenTerrain()
    #background_tiles = cimgt.OSM()

    crs_proj4 = crsOut.proj4_init
    country_prj=country.to_crs(crs_proj4)

    # this is just to extract the extent in projected coordinates
    # to use when testing using the ax.set_extent
    bbox=country_prj.bounds
    ymin = bbox['miny'].values[0]
    xmin = bbox['minx'].values[0]
    ymax = bbox['maxy'].values[0]
    xmax = bbox['maxx'].values[0]

    # define the basic matplotlib environment
    ax = plt.axes(projection=crsOut)

    # We can either set the axis extent OR plot the selected country
    # In both cases we get the same shift/stretch in the background image
    # However if we comment both lines, then everything plots
    # just fine at any zoom factor, but we get a map with the wrong limits

    #ax.set_extent([xmin, xmax, ymin, ymax], crs=crsOut)
    country_prj.plot(ax=ax, facecolor='forestgreen', edgecolor='0.5', alpha=0.4)

    # Also if we plot the entire cou dataset (without setting an extent), then there is no stretch
    #cou.to_crs(crs_proj4).plot(ax=ax, facecolor='forestgreen', edgecolor='0.5', alpha=0.4)

    # The common functionality vector data will always follow the crs setting defined
    # and align with the vectors of the shapefile
    ax.coastlines(color='red', linewidth=0.3)
    ax.add_feature(cfeature.BORDERS, edgecolor='blue', linewidth=0.2)

    # And here we add the background
    ax.add_image(background_tiles,zoom)

 if __name__ == '__main__':
    # Setting parameters and loading the data

    # Here we set the zoom level for the background image.
    # As we increase the zoom level, then the background stretches
    # more and more away from the -180/180 meridian:
    #       -stretches west when the central meridian of the projection is positive
    #       -stretches east when the central meridian of the projection is negative
    zoom=3

    # Ids of three countries that span across the -180/180 meridian
    # for which specific projection parameters are defines
    #id='NZL' # New Zealand
    #id = 'RUS'  # Russian Federation
    id='USA' # USA

    # Load any global country dataset with the ISO 3166-1 alpha-3 code
    cou = gp.read_file('../world/country.shp')

    # Select the country
    # NOTE: Change column index to the name of the attribute with the ISO 3166-1 alpha-3
    #       code in your country shapefile...
    country=cou.loc[cou['GMI_CNTRY'].isin([id])]

    # Projection parameters:
    #   [central meridian, latitude of true scale, minimum latitude, maximum latitude]
    projparam={
        'NZL':[174.885,9.081,-52.578,-34.415],
        'RUS':[108.815,20.328,41.190,81.852],
        'USA':[-127.245,26.213,-18.920,71.352]
    }

    cmerid=projparam[id][0]
    true_scale=projparam[id][1]
    min_lat=projparam[id][2]
    max_lat=projparam[id][3]

    # This is the projection definition used
    crsOut=ccrs.Mercator(central_longitude=cmerid, latitude_true_scale=true_scale, min_latitude=min_lat, max_latitude=max_lat)

    doPlot()
    plt.show()

```

#### Traceback 

```

```

<details>
 <summary>Full environment definition</summary>

### Operating system
Ubuntu Server 16.04 LTS

### Cartopy version
cartopy                   0.15.1

### conda list

```
# packages in environment at /xxx/xxx/xxx:
#
# Name                    Version                   Build  Channel
alabaster                 0.7.10                   py36_1    conda-forge
altair                    1.2.1                      py_0    conda-forge
asn1crypto                0.22.0                   py36_0    conda-forge
babel                     2.5.1                    py36_0    conda-forge
backports                 1.0                      py36_1    conda-forge
backports.functools_lru_cache 1.4                      py36_1    conda-forge
basemap                   1.1.0                    py36_3    conda-forge
basemap-data-hires        1.1.0                         0    conda-forge
biggus                    0.15.0                   py36_0    conda-forge
bleach                    2.0.0                      py_1    conda-forge
blinker                   1.4                        py_0    conda-forge
bokeh                     0.12.13                  py36_0    conda-forge
bottleneck                1.2.1                    py36_1    conda-forge
branca                    0.2.0                      py_1    conda-forge
bzip2                     1.0.6                h9a117a8_4  
ca-certificates           2018.1.18                     0    conda-forge
cairo                     1.14.12              h77bcde2_0  
cartopy                   0.15.1                   py36_5    conda-forge
certifi                   2018.1.18                py36_0    conda-forge
cf_units                  1.2.0                    py36_0    conda-forge
cffi                      1.11.2                   py36_0    conda-forge
chardet                   3.0.4                    py36_0    conda-forge
click                     6.7                        py_1    conda-forge
click-plugins             1.0.3                    py36_0    conda-forge
cligj                     0.4.0                    py36_0    conda-forge
cloudpickle               0.5.2                      py_0    conda-forge
cryptography              2.1.4                    py36_0    conda-forge
curl                      7.55.1                        0    conda-forge
cycler                    0.10.0                   py36_0    conda-forge
dask                      0.16.1                     py_0    conda-forge
dask-core                 0.16.1                     py_0    conda-forge
dbus                      1.10.22                       0    conda-forge
decorator                 4.1.2                    py36_0    conda-forge
descartes                 1.1.0                    py36_0    conda-forge
distributed               1.20.2                   py36_0    conda-forge
Django                    2.1.dev20180213191240           <pip>
docutils                  0.14                     py36_0    conda-forge
entrypoints               0.2.3                    py36_1    conda-forge
expat                     2.2.5                         0    conda-forge
ffmpeg                    3.4                  h7264315_0  
fiona                     1.7.9                    py36_0    conda-forge
flake8                    3.5.0                    py36_0    conda-forge
folium                    0.5.0                      py_0    conda-forge
fontconfig                2.12.6                        0    conda-forge
freetype                  2.8.1                         0    conda-forge
freexl                    1.0.4                         0    conda-forge
gdal                      2.1.4                    py36_2    conda-forge
geopandas                 0.3.0                    py36_0    conda-forge
geos                      3.6.2                         1    conda-forge
geotiff                   1.4.2                         1    conda-forge
gettext                   0.19.7                        1    conda-forge
giflib                    5.1.4                         0    conda-forge
glib                      2.53.6               h5d9569c_2  
gmp                       6.1.2                         0    conda-forge
gpxpy                     1.1.2                    py36_0    conda-forge
gst-plugins-base          1.8.0                         0    conda-forge
gstreamer                 1.8.0                         1    conda-forge
h5netcdf                  0.5.0                      py_0    conda-forge
h5py                      2.7.1                    py36_2    conda-forge
hdf4                      4.2.13                        0    conda-forge
hdf5                      1.10.1                        1    conda-forge
heapdict                  1.0.0                    py36_0    conda-forge
holoviews                 1.9.2                    py36_0    conda-forge
html5lib                  1.0.1                      py_0    conda-forge
icu                       58.2                          0    conda-forge
idna                      2.6                      py36_1    conda-forge
imagesize                 0.7.1                    py36_0    conda-forge
intel-openmp              2018.0.0             hc7b2577_8  
ipykernel                 4.8.0                    py36_0    conda-forge
ipython                   6.2.1                    py36_1    conda-forge
ipython_genutils          0.2.0                    py36_0    conda-forge
ipywidgets                7.1.1                    py36_0    conda-forge
iris                      1.13.0                   py36_1    conda-forge
jedi                      0.11.1                   py36_0    conda-forge
jenkspy                   0.1.4                     <pip>
jinja2                    2.10                     py36_0    conda-forge
jpeg                      9b                            2    conda-forge
json-c                    0.12.1                        0    conda-forge
jsonschema                2.6.0                    py36_1    conda-forge
jupyter                   1.0.0                    py36_0    conda-forge
jupyter_client            5.2.2                    py36_0    conda-forge
jupyter_console           5.2.0                    py36_0    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
kealib                    1.4.7                         4    conda-forge
krb5                      1.14.2                        0    conda-forge
libboost                  1.65.1               habcd387_4  
libcurl                   7.58.0               h1ad7b7a_0  
libdap4                   3.18.3                        2    conda-forge
libffi                    3.2.1                         3    conda-forge
libgcc-ng                 7.2.0                h7cc24e2_2  
libgdal                   2.1.4                         9    conda-forge
libgfortran               3.0.0                         1  
libgfortran-ng            7.2.0                h9f7466a_2  
libiconv                  1.15                          0    conda-forge
libkml                    1.3.0                h9d32c78_3  
libmo_unpack              3.0                           1    conda-forge
libnetcdf                 4.4.1.1                      10    conda-forge
libopus                   1.2.1                hb9ed12e_0  
libpng                    1.6.34                        0    conda-forge
libpq                     9.6.3                         0    conda-forge
libsodium                 1.0.15                        1    conda-forge
libspatialindex           1.8.5                         1    conda-forge
libspatialite             4.3.0a                       19    conda-forge
libssh2                   1.8.0                         2    conda-forge
libstdcxx-ng              7.2.0                h7a57d05_2  
libtiff                   4.0.9                         0    conda-forge
libuuid                   1.0.3                h1bed415_2  
libvpx                    1.6.1                h888fd40_0  
libxcb                    1.12                          1    conda-forge
libxml2                   2.9.7                         0    conda-forge
libxslt                   1.1.32                        0    conda-forge
locket                    0.2.0                    py36_1    conda-forge
lxml                      4.1.1                    py36_0    conda-forge
markupsafe                1.0                      py36_0    conda-forge
matplotlib                2.1.2                    py36_0    conda-forge
mccabe                    0.6.1                    py36_0    conda-forge
mistune                   0.8.3                      py_0    conda-forge
mkl                       2018.0.1             h19d6760_4  
mo_pack                   0.2.0                    py36_1    conda-forge
mplleaflet                0.0.5                    py36_1    conda-forge
msgpack-python            0.5.1                    py36_0    conda-forge
munch                     2.2.0                    py36_0    conda-forge
nbconvert                 5.3.1                      py_1    conda-forge
nbformat                  4.4.0                    py36_0    conda-forge
nc_time_axis              1.0.2                    py36_0    conda-forge
ncurses                   5.9                          10    conda-forge
netcdf4                   1.3.1                    py36_1    conda-forge
notebook                  5.4.0                    py36_0    conda-forge
numpy                     1.14.0           py36h3dfced4_1  
oauthlib                  2.0.6                      py_0    conda-forge
olefile                   0.44                     py36_0    conda-forge
openjpeg                  2.1.2                         3    conda-forge
openssl                   1.0.2n                        0    conda-forge
owslib                    0.16.0                     py_0    conda-forge
pandas                    0.22.0                   py36_0    conda-forge
pandoc                    2.1.1                         0    conda-forge
pandocfilters             1.4.1                    py36_0    conda-forge
param                     1.5.1                    py36_0    conda-forge
parso                     0.1.1                      py_0    conda-forge
partd                     0.3.8                    py36_0    conda-forge
pcre                      8.39                          0    conda-forge
pexpect                   4.3.1                    py36_0    conda-forge
pickleshare               0.7.4                    py36_0    conda-forge
pillow                    5.0.0                    py36_0    conda-forge
pip                       9.0.1                    py36_1    conda-forge
pixman                    0.34.0               hceecf20_3  
poppler-data              0.4.8                hf2eda46_0  
proj4                     4.9.3                         5    conda-forge
prompt_toolkit            1.0.15                   py36_0    conda-forge
psutil                    5.4.0                    py36_0    conda-forge
psycopg2                  2.7.4                    py36_0    conda-forge
ptyprocess                0.5.2                    py36_0    conda-forge
pycodestyle               2.3.1                    py36_0    conda-forge
pycparser                 2.18                     py36_0    conda-forge
pyepsg                    0.3.2                    py36_0    conda-forge
pyflakes                  1.6.0                    py36_0    conda-forge
pygments                  2.2.0                    py36_0    conda-forge
pyjwt                     1.5.3                      py_0    conda-forge
pyke                      1.1.1                    py36_0    conda-forge
pyopenssl                 17.4.0                   py36_0    conda-forge
pyparsing                 2.2.0                    py36_0    conda-forge
pyproj                    1.9.5.1                  py36_0    conda-forge
pyqt                      5.6.0                    py36_4    conda-forge
pysal                     1.14.3                   py36_0    conda-forge
pyshp                     1.2.12                     py_0    conda-forge
pysocks                   1.6.8                    py36_1    conda-forge
python                    3.6.4                         0    conda-forge
python-dateutil           2.6.1                    py36_0    conda-forge
pytz                      2017.3                     py_2    conda-forge
pyyaml                    3.12                     py36_1    conda-forge
pyzmq                     16.0.2                   py36_3    conda-forge
qt                        5.6.2                         7    conda-forge
qtconsole                 4.3.1                    py36_0    conda-forge
readline                  7.0                           0    conda-forge
reportlab                 3.4.0                     <pip>
requests                  2.18.4                   py36_1    conda-forge
requests-oauthlib         0.8.0                    py36_1    conda-forge
rtree                     0.8.3                    py36_0    conda-forge
scipy                     1.0.0            py36hbf646e7_0  
seawater                  3.3.4                    py36_0    conda-forge
send2trash                1.4.2                      py_0    conda-forge
setuptools                38.4.0                   py36_0    conda-forge
shapely                   1.6.4                    py36_0    conda-forge
simplegeneric             0.8.1                    py36_0    conda-forge
sip                       4.18                     py36_1    conda-forge
six                       1.11.0                   py36_1    conda-forge
snowballstemmer           1.2.1                    py36_0    conda-forge
sortedcontainers          1.5.7                    py36_0    conda-forge
sphinx                    1.6.7                    py36_0    conda-forge
sphinxcontrib-websupport  1.0.1                    py36_0    conda-forge
sqlalchemy                1.2.1                    py36_0    conda-forge
sqlite                    3.20.1                        2    conda-forge
tblib                     1.3.2                    py36_0    conda-forge
terminado                 0.8.1                    py36_0    conda-forge
testpath                  0.3.1                    py36_0    conda-forge
tk                        8.6.7                         0    conda-forge
toolz                     0.8.2                      py_2    conda-forge
tornado                   4.5.3                    py36_0    conda-forge
traitlets                 4.3.2                    py36_0    conda-forge
tweepy                    3.5.0                    py36_0    conda-forge
typing                    3.6.2                    py36_0    conda-forge
udunits2                  2.2.25                        2    conda-forge
urllib3                   1.22                     py36_0    conda-forge
util-linux                2.21                          0  
vega                      0.4.4                    py36_1    conda-forge
vincent                   0.4.4                    py36_0    conda-forge
wcwidth                   0.1.7                    py36_0    conda-forge
webencodings              0.5                      py36_0    conda-forge
wheel                     0.30.0                   py36_2    conda-forge
widgetsnbextension        3.1.3                    py36_0    conda-forge
xarray                    0.10.0                   py36_0    conda-forge
xerces-c                  3.2.0                         0    conda-forge
xorg-libxau               1.0.8                         3    conda-forge
xorg-libxdmcp             1.1.2                         3    conda-forge
xz                        5.2.3                         0    conda-forge
yaml                      0.1.6                         0    conda-forge
zeromq                    4.2.1                         1    conda-forge
zict                      0.1.3                      py_0    conda-forge
zlib                      1.2.11                        0    conda-forge
```

### pip list

```
```

</details>

**Comment by pelson** [comment]
Apologies for not getting back sooner @fabio1299.
I've just spent around 5 minutes trying to dig into the code, and unfortunately it is still way too complex to make this easy to reproduce - the use of globals makes the whole thing quite hard to follow, and I don't have access to your data file.

Really, I'm looking for a 4 line example to quickly see the problem. If there is a bug, I can then lift that example and use it as a test + then fix it. If there is a problem with the example itself, 4 lines is normally about right to make it quick and easy to see the problem.

Are you able to boil the problem down a little please @fabio1299? 

Thanks!

**Comment by pelson** [comment]
Just to add to this, the following code works fine:

```
from matplotlib import pyplot as plt
import cartopy.crs as ccrs
import cartopy.io.img_tiles as cimgt

ax = plt.axes(projection=ccrs.Mercator(central_longitude=180))

background_tiles = cimgt.StamenTerrain()
ax.add_image(background_tiles, 2)
ax.coastlines()
plt.show()
```

This suggests that there is probably an issue with the setup of your CRSs. Where I'd look first: when you're adding your data, are you setting the CRS correctly?

**Comment by QuLogic** [comment]
Closing as there has been no response in a while.
